ambry-network/src/main/java/com.github.ambry.network/BlockingChannelConnectionPool.java

```diff
@@ -223,6 +223,11 @@ public void destroyBlockingChannel(BlockingChannel blockingChannel) {
       synchronized (lock) {
         // decrement the number of connections to the host and port. we were not able to maintain the count
         numberOfConnections.decrementAndGet();
+        // at this point we are good to clean up the available connections since re-creation failed
+          while(!blockingChannelAvailableConnections.isEmpty()){
```

Comment:
There are a few locks in this class and the purposes are not immediately clear. Moreover, it seems that you can end up with a case where concurrent threads remove from the `blockingChannelAvailableConnections` at the same time. 

It is safer to do the following: 

```java
do {
  BlockingChannel channel = blockingChannelAvailableConnections.poll();
  if (channel == null) {
    break;
  }
  channel.disconnect();
  numberOfConnections.decrementAndGet();
} while (true);
```